addons/website_sale/controllers/main.py

```diff
@@ -1102,6 +1108,129 @@ def address(self, **kw):
+    def process_express_checkout(self, **post):
+        """ Records the partner information on the order when using express checkout flow.
+
+        Depending on whether the partner is registered and logged in, either creates a new partner
+        or uses an existing one that matches all received data.
+
+        :param dict post: a dictionary containing all the billing information sent by the express
+                          payment provider.
+        :return int: The order's partner id.
+        """
+        order = request.website.sale_get_order()
+
+        billing_address = post['billing_address']
+        # Update the partner with all the information
+        self._include_country_and_state_in_address(billing_address)
+
+        if order.partner_id.is_public:
+            self._convert_country_and_state_records_into_int(billing_address)
+            # If the partner doesn't exist yet, we create it.
+            billing_partner_id = self._create_or_edit_partner(
+                dict(billing_address, **{'type': 'invoice'})
+            )
+            order.partner_id = billing_partner_id
+            order.message_partner_ids = [Command.set([billing_partner_id])]
+        # If the partner exists, we use the address we already know.
+        elif any(billing_address[k] != order.partner_invoice_id[k] for k in billing_address):
+            billing_partners_sudo = request.env['res.partner'].with_context(
+                show_address=1
+            ).sudo().search([
+                ('id', 'child_of', order.partner_id.commercial_partner_id.ids),
+            ])
+            order.partner_invoice_id = False  # Remove the predefined invoice partner id
+            for billing_partner_sudo in billing_partners_sudo:
+                # Check if a child partner doesn't already exist with the same informations. The
+                # phone isn't always checked because it isn't sent in shipping information with
+                # Google Pay.
+                if all(billing_address[k] == billing_partner_sudo[k]for k in billing_address):
+                    order.partner_invoice_id = billing_partner_sudo.id
+                    break
+            if not order.partner_invoice_id:
+                self._convert_country_and_state_records_into_int(billing_address)
+                billing_partner_id = self._create_or_edit_partner(
+                    dict(billing_address, **{'type': 'invoice', 'parent_id': order.partner_id.id})
+                )
+                order.partner_invoice_id = billing_partner_id
+
+        # In a non-express flow, `sale_last_order_id` would be added in the session before the
+        # payment. As we skip all the steps with the express checkout, `sale_last_order_id` must be
+        # assigned to ensure the right behavior from `shop_payment_confirmation()`.
+        request.session['sale_last_order_id'] = order.id
+
+        return order.partner_id.id
+
+    def _include_country_and_state_in_address(self, address):
```

Comment:
What happens if we don't have any records matching country/state?

Reply:
The record included is empty and the partner has no country/state. No error is thrown nor logged. It's unlikely to happen because the country and the state are a selection in the Express checkout form managed by Google and Apple, and the records in Odoo can't be deleted or archived.